internal/pkg/util/env/create.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2021 Apptainer a Series of LF Projects LLC
+//   For website terms of use, trademark policy, privacy policy and other
+//   project policies see https://lfprojects.org/policies
+// Copyright (c) 2018-2020, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package env
+
+import (
+	"strings"
+
+	"github.com/apptainer/apptainer/internal/pkg/runtime/engine/config/oci/generate"
+	"github.com/apptainer/apptainer/pkg/sylog"
+)
+
+var alwaysPassKeys = map[string]struct{}{
+	"TERM":        {},
+	"http_proxy":  {},
+	"HTTP_PROXY":  {},
+	"https_proxy": {},
+	"HTTPS_PROXY": {},
+	"no_proxy":    {},
+	"NO_PROXY":    {},
+	"all_proxy":   {},
+	"ALL_PROXY":   {},
+	"ftp_proxy":   {},
+	"FTP_PROXY":   {},
+}
+
+// boolean value defines if the variable could be overridden
+// with the APPTAINERENV_ or SINGULARITYENV_ variant.
+var alwaysOmitKeys = map[string]bool{
+	"HOME":                false,
+	"PATH":                false,
+	"APPTAINER_SHELL":     false,
+	"APPTAINER_APPNAME":   false,
+	"SINGULARITY_SHELL":   false,
+	"SINGULARITY_APPNAME": false,
+	"LD_LIBRARY_PATH":     true,
+}
+
+// overridesForContainerEnv sets all environment variables which have overrides.
+func overridesForContainerEnv(g *generate.Generator, hostEnvs []string) map[string]string {
+	envKeys := make(map[string]string)
+	for _, prefix := range ApptainerEnvPrefixes {
+		for _, env := range hostEnvs {
+			if strings.HasPrefix(env, prefix) {
+				e := strings.SplitN(env, "=", 2)
+				if len(e) != 2 {
+					sylog.Verbosef("Can't process override environment variable %s", env)
+				} else {
+					key := e[0][len(prefix):]
+					switch key {
+					case "PREPEND_PATH":
+						envKeys["SING_USER_DEFINED_PREPEND_PATH"] = e[1]
+						envKeys["APPTAINER_USER_DEFINED_PREPEND_PATH"] = e[1]
+					case "APPEND_PATH":
+						envKeys["SING_USER_DEFINED_APPEND_PATH"] = e[1]
+						envKeys["APPTAINER_USER_DEFINED_APPEND_PATH"] = e[1]
+					case "PATH":
+						envKeys["SING_USER_DEFINED_PATH"] = e[1]
+						envKeys["APPTAINER_USER_DEFINED_PATH"] = e[1]
+					default:
+						if key != "" {
+							if permitted, ok := alwaysOmitKeys[key]; ok && !permitted {
+								sylog.Warningf("Overriding %s environment variable with %s is not permitted", key, e[0])
+								continue
+							}
+							if value, ok := envKeys[key]; ok {
+								sylog.Verbosef("Skipping environment variable [%s=%s], %s is already overridden with [%s]", e[0], e[1], key, value)
```

Comment:
Based on the discussion in #7, it looks like this should be `Warningf` so users get output at the default log level

Reply:
@ikaneshiro addressed with a common function for all environment variable overrides